app/controllers/incomeSources/add/AddIncomeSourceStartDateController.scala

```diff
@@ -173,26 +172,27 @@ class AddIncomeSourceStartDateController @Inject()(authenticate: AuthenticationP
 
   private def getFilledForm(form: Form[DateFormElement],
                             incomeSourceType: IncomeSourceType,
-                            isUpdate: Boolean)(implicit user: MtdItUser[_]): Either[Throwable, Form[DateFormElement]] = {
-
-    val maybeStartDate = getStartDate(incomeSourceType)
+                            isUpdate: Boolean)(implicit user: MtdItUser[_]): Future[Either[Throwable, Form[DateFormElement]]] = {
 
-    (maybeStartDate, isUpdate) match {
-      case (Some(date), true) if Try(LocalDate.parse(date)).toOption.isDefined =>
+    getStartDate(incomeSourceType) map {
+      case Some(date) if isUpdate && Try(LocalDate.parse(date)).toOption.isDefined =>
         Right(
           form.fill(
             DateFormElement(
               LocalDate.parse(date)
             )
           )
         )
-      case (Some(date), true) => Left(new Error(s"Could not parse $date as a LocalDate"))
+      case Some(date) if isUpdate => Left(new Error(s"Could not parse $date as a LocalDate"))
       case _ => Right(form)
     }
   }
 
-  private def getStartDate(incomeSourceType: IncomeSourceType)(implicit user: MtdItUser[_]): Option[String] = {
-    user.session.get(incomeSourceType.startDateSessionKey)
+  private def getStartDate(incomeSourceType: IncomeSourceType)(implicit user: MtdItUser[_]): Future[Option[String]] = {
```

Comment:
if you can change signature of this method to Future[Eithe[Throwable,LocalDate]]
* extract startSessionDateKey value and convert to LocalDate
* return errors if any